### ISSUE TITLE
Auto render mode is not working as expected

### ISSUE THREAD

**Issue by yasmoradi** [danroth27/Net8BlazorAuto#4]
I made changes to the project that you can see in [this commit](https://github.com/ysmoradi/Net8BlazorAuto/commit/1278774e9fa36d672d9e006f3ca65a92ba45f132).

My questions:

1- Why does the state of the two inputs in the Layout get lost when switching between two pages that are in auto render mode?
First, go to the counter, type something in the two inputs, then go to counter2 and you'll see the input values are lost!

2- Why, when I move from counter to counter2, and considering that the current count is printed in the browser console, showing that the Blazor Web Assembly has successfully initialized and is working, when I shut down the server or disconnect the network, the whole site stops working? I expected that after Blazor WebAssembly initialization, the web socket would disconnect, the server resources would be freed, and my client would no longer depend on the server.

3- **The second time** I refresh the page, I set the Network to 3G for testing. Why, even though the download of the dotnet native **_js_** takes more than 500 milliseconds, doesn't the site start with Blazor server and keeps waiting for Blazor wasm?

4- Why does the Blazor WebAssembly initialization, which is time-consuming on mid/low-end phones, start only after moving to the second page? Even if we wait on the first page for an hour, we don't take advantage of this time to run Blazor WebAssembly.

**Current behavior is right now:**
The first time it's always Blazor Server, subsequent times it's always Blazor WebAssembly, regardless of how long it takes.
If it starts with Blazor Server, the Blazor WebAssembly initialization doesn't happen while the user is on the first page, and it gets postponed without any advantage to the time the user goes to the second page, which I think is late.
Also, the state present in the layout is lost during transitions.

**Ideal conditions for Auto Render Mode:**

1- If the execution of Blazor WebAssembly is quick, use it. Otherwise, use fallback to Blazor server. It seems that from the second time onwards, regardless of how long it takes for Blazor WebAssembly to execute, the program starts with Blazor WebAssembly!

2- If the program starts with Blazor server, Blazor WebAssembly should begin its initialization in the background without waiting for a route change or anything else, especially since it's considerably slow on weak devices.

3- Whenever Blazor WebAssembly starts working and takes control of the UI, the web socket should disconnect, and the client should no longer be dependent on the server in this regard.

4- When switching between two pages that are both in auto mode, the state in the layout should not be lost. Now, even if both execute with WebAssembly, switching between them still results in the loss of the layout state.

**Comment by danroth27** [comment]
> 1- Why does the state of the two inputs in the Layout get lost when switching between two pages that are in auto render mode? First, go to the counter, type something in the two inputs, then go to counter2 and you'll see the input values are lost!

In this sample page navigations are handled using server-side rendering, so no state is preserved across pages. If you wanted instead to make the entire app use the Auto render mode then that's certainly doable.

**Comment by danroth27** [comment]
> 2- Why, when I move from counter to counter2, and considering that the current count is printed in the browser console, showing that the Blazor Web Assembly has successfully initialized and is working, when I shut down the server or disconnect the network, the whole site stops working? I expected that after Blazor WebAssembly initialization, the web socket would disconnect, the server resources would be freed, and my client would no longer depend on the server.

The counter should still work, but the rest of the app is server-based in this sample.

**Comment by danroth27** [comment]
> 3- The second time I refresh the page, I set the Network to 3G for testing. Why, even though the download of the dotnet native js takes more than 500 milliseconds, doesn't the site start with Blazor server and keeps waiting for Blazor wasm?

Hmm...I'm not sure on this one. @MackinnonBuck?

**Comment by danroth27** [comment]
> 4- Why does the Blazor WebAssembly initialization, which is time-consuming on mid/low-end phones, start only after moving to the second page? Even if we wait on the first page for an hour, we don't take advantage of this time to run Blazor WebAssembly.

There's quite a bit of complexity to shift the state from the server to the client while the page loaded and being used.

**Comment by MackinnonBuck** [comment]
> > 3- The second time I refresh the page, I set the Network to 3G for testing. Why, even though the download of the dotnet native js takes more than 500 milliseconds, doesn't the site start with Blazor server and keeps waiting for Blazor wasm?
> 
> Hmm...I'm not sure on this one. @MackinnonBuck?

We need to load `blazor.boot.json` to determine if the WebAssembly resource hash has been seen before. That's how we know whether WebAssembly resources are cached, and loading `dotnet.js` is a prerequisite to that. After loading `blazor.boot.json`, if it's determined that remaining WebAssembly resources are not already cached, then Blazor Server is used. Otherwise, Blazor WebAssembly is used. There's a fallback timeout of 3 seconds for this loading process, so Blazor Server will get used if that timeout gets hit. Maybe we could make that timeout configurable.

> 1- Why does the state of the two inputs in the Layout get lost when switching between two pages that are in auto render mode?
First, go to the counter, type something in the two inputs, then go to counter2 and you'll see the input values are lost!

To expand on @danroth27's point, the state rendered by the server is what gets displayed on the page. Since the layout renders an input with an empty value, that's what shows up on the page after navigation.

However, it does seem like a valid request to somehow preserve user-entered input state across enhanced navigations. I remember @captainsafia and I encountered the state-erasing behavior when prototyping the hot reload implementation, and it did seem like a bug at first glance. Maybe we should consider adding APIs or conventions to preserve user-entered state? cc @SteveSandersonMS

> 2- Why, when I move from counter to counter2, and considering that the current count is printed in the browser console, showing that the Blazor Web Assembly has successfully initialized and is working, when I shut down the server or disconnect the network, the whole site stops working? I expected that after Blazor WebAssembly initialization, the web socket would disconnect, the server resources would be freed, and my client would no longer depend on the server.

It seems like part of what you're asking for is being implemented in https://github.com/dotnet/aspnetcore/pull/50170. But as @danroth27 pointed out, there's not currently a way to dynamically change the render mode of a component without losing state. So, for example, if an Auto component gets initialized using Blazor Server, it will continue to use Blazor Server. As of https://github.com/dotnet/aspnetcore/pull/50170, the circuit will only shut down after that component (and all other Server components) get removed from the page.

> 4- Why does the Blazor WebAssembly initialization, which is time-consuming on mid/low-end phones, start only after moving to the second page? Even if we wait on the first page for an hour, we don't take advantage of this time to run Blazor WebAssembly.

Blazor WebAssembly doesn't start until a Blazor WebAssembly component gets added to the page. What if the customer never navigates to the second page? Would you still want to execute the resource-intensive initialization of Blazor WebAssembly?

That said - it's a valid feature request to want to start up Blazor WebAssembly on a whim rather than waiting for a Blazor WebAssembly component to get added to the page.

> It seems that from the second time onwards, regardless of how long it takes for Blazor WebAssembly to execute, the program starts with Blazor WebAssembly!

The auto mode implementation uses Blazor WebAssembly if its resources are already loaded/cached. It doesn't cancel the initialization of the WebAssembly runtime (or fall back on Blazor Server) if it happens to be taking a long time. However, it's a valid feature request to configure a timeout for Blazor WebAsembly startup.

> If the program starts with Blazor server, Blazor WebAssembly should begin its initialization in the background without waiting for a route change or anything else, especially since it's considerably slow on weak devices.

Again, this isn't always ideal because you can't know for sure that the customer is going to navigate to a page with Blazor WebAssembly components.

> Whenever Blazor WebAssembly starts working and takes control of the UI, the web socket should disconnect, and the client should no longer be dependent on the server in this regard.

Not currently possible due to the aforementioned state shifting problem.

@ysmoradi, many of your requests are valid. However, they will require more analysis, design, and feedback than what we are able to complete before the .NET 8 release. Please feel free to file issues in https://github.com/dotnet/aspnetcore with your feature requests and bug reports so the team can review/address them!

**Comment by SteveSandersonMS** [comment]
> However, it does seem like a valid request to somehow preserve user-entered input state across enhanced navigations

That sounds like a variation on this prototype/experiment I did recently: https://github.com/SteveSandersonMS/StatefulReconnection. This preserves form state across reconnections. The same technique could be used across navigations.

I didn't get a huge amount of feedback from the experiment so it's not clear to me how much demand there is for this. It's possible that having form fields in the *layout* part of a site (and expecting them to be preserved across page loads) is a very niche scenario. It would also be inconsistent with non-enhanced navigation to preserve the field state, and the primary goal of enhanced nav is to recreate the same semantics of non-enhanced nav but faster - with the one key exception of preserving interactive components.

> Why does the Blazor WebAssembly initialization, which is time-consuming on mid/low-end phones, start only after moving to the second page? Even if we wait on the first page for an hour, we don't take advantage of this time to run Blazor WebAssembly.

This point was raised a few times above. As @MackinnonBuck says we don't load WebAssembly until there's an indication the user would benefit from it. But the developer can control this: if you want to *always* load WebAssembly in the background, simply put an empty component in the layout with `WebAssembly` render mode. Then your app will always start loading the WebAssembly runtime immediately, even if `Auto` resolves as `Server` for other components (is that right, @MackinnonBuck?).

**Comment by danroth27** [comment]
Closing as I think all questions have been answered here.
